### ISSUE TITLE
MLPack C++ for DecisionTree based problem

### ISSUE THREAD

**Issue by saanumm82** [mlpack/mlpack#2138]
MLPack C++ for decision tree based problems - 

I would like to use MLPack C++ for below problem to forecast whether one will be able to play the game in the given weather conditions using ML Decision tree.

Here are the train and test data sets. Based on this, i want to build a decision tree to predict the data for test data

Train Dataset(train.csv)

Outlook Temperature Humidity    Windy   Play
Sunny   Hot     High    False   No
Rainy   Mild    High    False   Yes
Sunny   Cool    Normal  False   Yes
Overcast    Hot High    False   Yes
Rainy   Mild    High    False   Yes
Overcast    Hot Normal  False   Yes
Test Dataset(test.csv)

Id  Outlook Temperature Humidity    Windy
1   Sunny   Mild    Normal  True
2   Sunny   Mild    High    False
3   Overcast    Cool    Normal  True
4   Rainy   Mild    High    True
The predicated out put should be as given below.

Id  Play
1   Yes
2   Yes
3   No
4   No
I want to know can i use MlPack c++ for this kind of problems?  Please let me know how to solve above problem using MLPack  C++ DecisionTree class or any other class.

Appreciate sample code to solve above problem. Problem details you can find in - https://www.hackerearth.com/blog/developers/machine-learning-auto-evalaution/

**Comment by saanumm82** [comment]
I have written this code to solve the problem. But looks like it's not building DecisionTree so output is wrong. Please let me know why this code is not working.

#include <mlpack/core.hpp>
#include <mlpack/methods/neighbor_search/neighbor_search.hpp>
#include <mlpack/core.hpp>
#include <mlpack/methods/decision_tree/decision_tree.hpp>
#include <mlpack/methods/decision_tree/information_gain.hpp>
#include <mlpack/methods/decision_tree/gini_gain.hpp>
#include <mlpack/methods/decision_tree/random_dimension_select.hpp>
#include <mlpack/methods/decision_tree/multiple_random_dimension_select.hpp>


using namespace std;
using namespace mlpack;
// NeighborSearch and NearestNeighborSort 
using namespace mlpack::neighbor;
// ManhattanDistance 
using namespace mlpack::metric;
using namespace mlpack::tree;
using namespace mlpack::distribution;

int main()
{
  arma::mat dataset;
  arma::Row<size_t> labels;
  data::Load("vc2.csv", dataset);
  data::Load("vc2_labels.txt", labels);
  

      DecisionTree<GiniGain, BestBinaryNumericSplit, AllCategoricalSplit,
                AllDimensionSelect, double, true> d(dataset, labels, 2, 1);

  cout<<"child:"<<d.NumChildren()<<endl;
  cout<<"dt.NumClasses():"<<d.NumClasses()<<endl;

  arma::mat testData;
  data::Load("vc2_test.csv", testData);
  arma::Row<size_t> predictions;
  d.Classify(testData, predictions);
  predictions.print("prediction:");
  return 0;
}



vc2.csv file has below dataset

Sunny,Hot,High,Weak
Sunny,Hot,High,Strong
Overcast,Hot,High,Weak
Rain,Mild,High,Weak
Rain,Cool,Normal,Weak
Rain,Cool,Normal,Strong
Overcast,Cool,Normal,Strong
Sunny,Mild,High,Weak
Sunny,Cool,Normal,Weak
Rain,Mild,Normal,Weak
Sunny,Mild,Normal,Strong
Overcast,Mild,High,Strong
Overcast,Hot,Normal,Weak
Rain,Mild,High,Strong

and 
vc2_labels.txt has following content ( here 0 means No and 1 means Yes )

0
0
1
1
1
0
1
0
1
1
1
1
1
0

and vc2_test.csv file has 

Overcast,Hot,High,Weak.


With these data, prediction is wrong.

**Comment by adithya-tp** [comment]
Hey @saanumm82 ! From what I can see from the documentation, it seems like "string" categorical data is not supported by mlpack at the moment with the exception of "mlpack_hoeffding_tree".
(refer here: [https://www.mlpack.org/doc/stable/doxygen/formatdoc.html](url) [see subsection **Categorical features and command line programs** - last line])

However, we can work around this by mapping the categorical variables in each column to numbers. Essentially performing "Label Encoding" on each column in the dataset.

So the training data set that originally looks like this (I've chosen these rows just as an example):

>Sunny,Hot,High,Weak
Sunny,Hot,High,Strong
Rain,Mild,High,Weak
Rain,Cool,Normal,Weak
Rain,Cool,Normal,Strong
Overcast,Cool,Normal,Strong
Sunny,Mild,High,Weak
Sunny,Cool,Normal,Weak
Rain,Mild,Normal,Weak
Sunny,Mild,Normal,Strong
Overcast,Mild,High,Strong
Overcast,Hot,Normal,Weak
Rain,Mild,High,Strong

would now look like as follows **(train.csv):** 

>2,1,0,1
2,1,0,0
1,2,0,1
1,0,1,1
1,0,1,0
0,0,1,0
2,2,0,1
2,0,1,1
1,2,1,1
2,2,1,0
0,2,0,0
0,1,1,1
1,2,0,0

Along the same lines, if the test case is: `Overcast,Hot,High,Weak` it gets converted to `0,1,0,1` (**test.csv**).

Of course, the numeric labels remain the same(**labels.txt**). For this particular example, the training labels are:

> 0
0
1
1
0
1
0
1
1
1
1
1
0

Now to fit a decision tree on this training data, I've modified your code just a bit. Here the only major change is the `d.Train(dataset, labels, 2, 1);` to fit the **DecisionTree** classifier onto the data.

What follows is the code that I used to solve your problem using mlpack:


        #include <iostream>
        #include <mlpack/core.hpp>
        #include <mlpack/methods/decision_tree/decision_tree.hpp>
        #include <mlpack/methods/decision_tree/information_gain.hpp>
        #include <mlpack/methods/decision_tree/gini_gain.hpp>
        #include <mlpack/methods/decision_tree/random_dimension_select.hpp>
        #include <mlpack/methods/decision_tree/multiple_random_dimension_select.hpp>

        using namespace std;
        using namespace arma;
        using namespace mlpack;
        using namespace mlpack::metric;
        using namespace mlpack::tree;
        using namespace mlpack::distribution;

        int main(int argc, char** argv) {
            /*
                When calling the executable pass in the file names for "numeric train" file and "numeric labels" file.
                
                Generic example: ./<executable_name> <numeric_training_file> <numeric_labels_file>
                Specific example: ./a.out train_numeric.csv labels.txt
            */

            // to check if the training and label files are being passed in.
            if(argc < 3) throw runtime_error("error: [bad syntax] please pass a dataset for the decision tree");
            arma::mat dataset;
            arma::Row<size_t> labels;

            data::Load(argv[1], dataset);
            data::Load(argv[2], labels);

            /*
                Printing out the training data and labeling data as a sanity check.
                I guess we can afford to do this since it is a small enough dataset.
            */
        
            dataset.print("Training Data: ");
            labels.print("Labels For Training Data: ");

            // creating a tree object
            DecisionTree<GiniGain, BestBinaryNumericSplit, AllCategoricalSplit, AllDimensionSelect, double> d;
            // fitting the tree on the dataset (dataset, labels, numClasses, minimumLeafSize).
            d.Train(dataset, labels, 2, 1);

            cout << "Number of children: "<< d.NumChildren() << "\n";
            cout << "dt.NumClasses():" << d.NumClasses() << "\n";

            arma::mat testData;
            data::Load("test_numeric.csv", testData);
            testData.print("Test Data: ");
            arma::Row<size_t> predictions;
            d.Classify(testData, predictions);
            predictions.print("prediction:");
            return 0;
        }

`

I have included comments to make my code clear. However, feel free to ask any doubts you have about the same. I'll try my best to help you out.

I compiled my code using:
```g++ dt.cpp -o dt  `pkg-config --cflags --libs mlpack armadillo`  -fopenmp```
and ran the executable using:
`./dt train_numeric.csv labels.txt`

Here is the output i got with the right prediction ( = 1) for `Overcast,Hot,High,Weak`:
(each column is one training point due to the way mlpack loads matrices - hence we have 13 columns for each of the 13 training examples).
> Training Data: 
   2.0000  2.0000   1.0000   1.0000   1.0000        0   2.0000   2.0000   1.0000   2.0000        0        0   1.0000
   1.0000   1.0000   2.0000        0        0        0   2.0000        0   2.0000   2.0000   2.0000   1.0000   2.0000
        0        0        0   1.0000   1.0000   1.0000        0   1.0000   1.0000   1.0000        0   1.0000        0
   1.0000        0   1.0000   1.0000        0        0   1.0000   1.0000   1.0000        0        0   1.0000        0
Labels For Training Data: 
        0        0        1        1        0        1        0        1        1        1        1        1        0
Number of children: 2
dt.NumClasses():2
Test Data: 
        0
   1.0000
        0
   1.0000
prediction:
        1

I hope this resolves your issue. Cheers!

**Comment by saanumm82** [comment]
Thanks, I have another doubt on MLPack C++. Do you know which companies are using MLPack C++ in their production code.

I want use it for productions code implementation. So wanted to know should I need to go for MLPack C++ or tensorflow.

My requirements is almost same as that of example mentioned in this email thread.

Please suggest.

**Comment by adithya-tp** [comment]
I'm not sure about that info. However, I'm quite certain you could use mlpack c++ in production for the purposes mentioned in this thread, since this is independent of which companies use mlpack.

Also if your initial issue has been resolved, please feel free to close this issue.

Glad I could help you out !

**Comment by saanumm82** [comment]
Thanks a lot for your help on this. The original issue is resolved hence closing this issue
